java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -51,191 +51,211 @@
+      }
+      SymbolicValue invocationTarget = parent.programState.peekValue(mit.arguments().size());
+      if (symbolicValues.contains(invocationTarget)) {
+        flowBuilder.add(location(parent, "..."));
+      }
+      int argIdx = correspondingArgumentIndex(symbolicValues, parent);
+      if (argIdx != -1) {
+        ExpressionTree argTree = mit.arguments().get(argIdx);
+        String message = String.format("Implies '%s' is %s.", SyntaxTreeNameFinder.getName(argTree), learnedConstraint.valueAsString());
+        flowBuilder.add(new JavaFileScannerContext.Location(message, argTree));
+      }
+      MethodYield selectedMethodYield = edge.child.selectedMethodYield(parent);
+      if (selectedMethodYield != null) {
+        Set<List<JavaFileScannerContext.Location>> xProcFlows = selectedMethodYield.flow(argIdx, domains);
+        // FIXME SONARJAVA-2076 consider all flows from yields
+        xProcFlows.stream().findFirst().ifPresent(f -> f.forEach(flowBuilder::add));
+      }
+      return flowBuilder.build();
+    }
 
-  private static List<SymbolicValue> argumentsUsedForMethodInvocation(ExplodedGraph.Node invocationNode, MethodInvocationTree mit) {
-    List<SymbolicValue> values = invocationNode.programState.peekValues(mit.arguments().size());
-    return Lists.reverse(values);
-  }
-  @Nullable
-  private Symbol addFlowFromLearnedSymbols(ExplodedGraph.Node currentNode, @Nullable Symbol trackSymbol) {
-    ExplodedGraph.Node parent = currentNode.parent();
-    if (trackSymbol == null || parent == null) {
-      return null;
+    private boolean isMethodInvocationNode(ExplodedGraph.Node node) {
+      // ProgramPoint#syntaxTree will not always return the correct tree, so we need to go to ProgramPoint#block directly
+      ProgramPoint pp = node.programPoint;
+      if (pp.i < pp.block.elements().size()) {
+        Tree tree = pp.block.elements().get(pp.i);
+        return tree.is(Tree.Kind.METHOD_INVOCATION);
+      }
+      return false;
     }
-    Optional<LearnedAssociation> learnedAssociation = currentNode.learnedAssociations()
-      .filter(lv -> lv.symbol().equals(trackSymbol))
-      .findFirst();
-    if (learnedAssociation.isPresent()) {
-      LearnedAssociation la = learnedAssociation.get();
-      for (Class<? extends Constraint> domain : domains) {
-        Constraint constraint = parent.programState.getConstraint(la.symbolicValue(), domain);
-        if(constraint != null) {
-          flow.add(location(parent, String.format("'%s' is assigned %s.", la.symbol().name(), constraint.valueAsString())));
+
+    private int correspondingArgumentIndex(Set<SymbolicValue> candidates, ExplodedGraph.Node invocationNode) {
+      MethodInvocationTree mit = (MethodInvocationTree) invocationNode.programPoint.syntaxTree();
+      List<SymbolicValue> arguments = argumentsUsedForMethodInvocation(invocationNode, mit);
+      // FIXME SONARJAVA-2076 whatif multiple candidates match some of the arguments?
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)